src/mlpack/methods/neighbor_search/neighbor_search.hpp

```diff
@@ -309,14 +412,21 @@ class NeighborSearch
   //! Return the number of node combination scores during the last search.
   size_t Scores() const { return scores; }
 
+  //! Access the search mode.
+  NeighborSearchMode SearchMode() const { return searchMode; }
+  //! Modify the search mode.
+  void SetSearchMode(const NeighborSearchMode mode);
```

Comment:
You're right; I thought there was a way to work around this problem but really there is not a good one.  What I really want to avoid is making an API change that we'll regret later (like `SetSearchMode()`).  So if we are going to break reverse compatibility in 3.0.0, then I would say maybe the easiest thing to do here is just add a `greedy` boolean parameter that works the same way as `singleMode` and `naive`, and use that for now.  Then, when we release 3.0.0, we'll remove all three of those and _only_ have `searchMode` in the way you've written here.  This would allow us to avoid the weird stuff we have to do for reverse compatibility.

I guess, if you agree with that, then we should save your commits for refactoring by opening another reverse-incompatible "3.0.0 PR" that will be merged in when we decide it is time for a release there.